#  Описание проекта

__Дано:__ крупный сервис для чтения книг по подписке.

__Задача:__ проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

# Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Содержание

1. [Открытие данных, первая информация о данных](#start)


2. [Выполнение заданий](#completing_tasks)

    2.1 [Посчитайте, сколько книг вышло после 1 января 2000 года](#task1)
    
    2.2 [Для каждой книги посчитайте количество обзоров и среднюю оценку](#task2)
    
    2.3 [Определите издательство, которое выпустило наибольшее число книг толще 50 страниц](#task3)
    
    2.4 [Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками](#task4)
    
    2.5 [Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.](#task5)
    
    
3. [Выводы](#conclusions)

<a id='start'></a>
## Открытие данных 

In [1]:
import pandas as pd
from sqlalchemy import create_engine

__Подключение к базе данных__

In [2]:
db_config = {'user': 'praktikum_student', 
'pwd': 'Sdf4$2;d-d30pp', 
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, 
'db': 'data-analyst-final-project-db'}

In [3]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

In [4]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 


__Таблица books__

In [5]:
query = '''
            SELECT * FROM books
        '''

In [6]:
books = pd.io.sql.read_sql(query, con = engine) 

In [7]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


__Таблица authors__

In [9]:
query = '''
            SELECT *
            FROM authors
        '''

In [10]:
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [11]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


__Таблица publishers__

In [12]:
query = '''
            SELECT * FROM publishers
        '''

In [13]:
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [14]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


__Таблица ratings__

In [15]:
query = '''
            SELECT * FROM ratings
        '''

In [16]:
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [17]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


__Таблица reviews__

In [18]:
query = '''
            SELECT * FROM reviews
        '''

In [19]:
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [20]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


<a id='completing_tasks'></a>
## Выполнение заданий 

<a id='task1'></a>
### Посчитайте, сколько книг вышло после 1 января 2000 года

In [21]:
query = '''
            SELECT COUNT(book_id) 
            FROM books
            WHERE publication_date::date > '2000-01-01'
        '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

В период после 1 января 2000 года вышло 819 книг

<a id='task2'></a>
## Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [22]:
query = '''
            WITH rat AS (SELECT book_id, AVG(rating) avg_rating 
                        FROM ratings 
                        GROUP BY book_id),
                rev AS (SELECT book_id, COUNT(review_id) review_count 
                         FROM reviews 
                         GROUP BY book_id)
            SELECT b.title, b.book_id, avg_rating, review_count
            FROM books AS b
            LEFT JOIN  rat ON b.book_id=rat.book_id
            LEFT JOIN  rev ON b.book_id=rev.book_id
            
        '''
pd.io.sql.read_sql(query, con = engine)

title  book_id  avg_rating  \
0            The Body in the Library (Miss Marple  #3)      652    4.500000   
1                                            Galápagos      273    4.500000   
2                             A Tree Grows in Brooklyn       51    4.250000   
3    Undaunted Courage: The Pioneering First Missio...      951    4.000000   
4                                          The Prophet      839    4.285714   
..                                                 ...      ...         ...   
995   The Cat in the Hat and Other Dr. Seuss Favorites      672    5.000000   
996    Anne Rice's The Vampire Lestat: A Graphic Novel       83    3.666667   
997                          Essential Tales and Poems      221    4.000000   
998                               Leonardo's Notebooks      387    4.000000   
999                            The Natural Way to Draw      808    3.000000   

     review_count  
0             2.0  
1             2.0  
2             5.0  
3             2.0  
4             4.0  
..            ...  
995           NaN  
996           NaN  
997           NaN  
998           NaN  
999           NaN  

[1000 rows x 4 columns]

Посчитали количество обзоров и среднюю оценку для каждой книги. На некоторые книги отсутствуют обзоры.

<a id='task3'></a>
### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [23]:
query = '''
            WITH b as (SELECT 
                              COUNT(book_id) AS books , publisher_id
                        FROM books
                        WHERE num_pages > 50
                        GROUP BY publisher_id)
            SELECT 
                   publisher,  books, p.publisher_id as publisher_id
            FROM b
            LEFT JOIN publishers AS p  ON b.publisher_id=p.publisher_id
            ORDER BY books DESC
            LIMIT 1
        '''
       
pd.io.sql.read_sql(query, con = engine)

publisher  books  publisher_id
0  Penguin Books     42           212

Издательство Penguin Books выпустило максимальное количество книг с количеством страниц больше 50

<a id='task4'></a>
### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [24]:
query = '''
            WITH rat AS (SELECT
                                book_id, COUNT(rating_id) AS count_rating, AVG(rating) AS avg_rating 

                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating_id)>50)
            SELECT 
                    author, AVG(avg_rating)
            FROM books AS b
            JOIN rat ON rat.book_id=b.book_id
            JOIN authors as a ON b.author_id=a.author_id
            GROUP BY author
            ORDER BY AVG(avg_rating) DESC
            LIMIT 1

        '''
       
pd.io.sql.read_sql(query, con = engine)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

`Joanne Kathleen Rowling`в соавторстве с иллюстратором `Mary GrandPré` автор с самой высокой средней оценкой книг из книг с больше 50-ю оценками.

<a id='task5'></a>
### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [25]:
query = ''' WITH a as(SELECT
                              COUNT(review_id) AS count,
                              username
                         FROM reviews 
                         GROUP BY username),
                 b as (SELECT
                              username, 
                              COUNT(rating_id)
                        FROM ratings
                        GROUP BY username
                        HAVING count(rating_id)>50)
            SELECT AVG(a.count)
            FROM a
            INNER JOIN b ON a.username=b.username
                         
  '''
pd.io.sql.read_sql(query, con = engine)                        

avg
0  24.333333

В среднем 24 обзора пишут пользователи, которые поставили больше 50 оценок

<a id='conclusions'></a>
## Выводы

В нашем распоряжении было 5 таблиц:

- books - содержит 1000 строк и 6 столбцов, содержит данные о книгах;
- authors - содержит 636 строк и 2 столбца, содержит данные об авторах;
- publishers - содержит 340 строк, 2 столбца, содержит данные об издательствах;
- rating - содержит 6456 строк, 4 столбца, содержит данные о пользовательских оценках;
- reviews - содержит 2793 строки, 4 столбца, содержит данные о пользовательских обзорах.

__В период после 1 января 2000 года вышло 819 книг__


__Посчитали количество обзоров и среднюю оценку__


__Издательство Penguin Books выпустило максимальное количество книг с количеством страниц больше 50__

Penguin Books — британское издательство, основанное в 1935 году в Лондоне сэром Алленом Лейном и его братьями Ричардом и Джоном. Главной заслугой издательства считают демократизацию книжного рынка, превратившую книгу из предмета роскоши в удовольствие, доступное массам.

__Joanne Kathleen Rowling в соавторстве с иллюстратором Mary GrandPré - автор с самой высокой средней оценкой книг из книг с больше 50-ю оценками__

Joanne Kathleen Rowling - автор серии книг о Гарри Поттере. Mary GrandPré - американский иллюстратор и писатель, наиболее известный своими иллюстрациями изданий книг о Гарри Поттере в США, Бразилии и России.

__В среднем 24 обзора пишут пользователи, которые поставили больше 50 оценок__
